react-components/src/components/NodeCacheProvider/AssetMappingCache.ts

```diff
@@ -0,0 +1,87 @@
+/*!
+ * Copyright 2023 Cognite AS
+ */
+
+import { type CogniteClient, type AssetMapping3D, type Node3D } from '@cognite/sdk';
+import { type ModelId, type ModelRevisionKey, type RevisionId } from './types';
+import { maxBy } from 'lodash';
+import assert from 'assert';
+
+export type NodeAssetMappingResult = { node?: Node3D; mappings: AssetMapping3D[] };
+
+export class AssetMappingCache {
+  private readonly _sdk: CogniteClient;
+
+  private readonly _modelToAssetMappings = new Map<ModelRevisionKey, AssetMapping3D[]>();
+
+  constructor(sdk: CogniteClient) {
+    this._sdk = sdk;
+  }
+
+  public async getAssetMappingsForLowestAncestor(
+    modelId: ModelId,
+    revisionId: RevisionId,
+    ancestors: Node3D[]
+  ): Promise<NodeAssetMappingResult> {
+    if (ancestors.length === 0) {
+      return { mappings: [] };
+    }
+
+    const searchTreeIndices = new Set(ancestors.map((ancestor) => ancestor.treeIndex));
+    const allModelMappings = await this.getAssetMappingsForModel(modelId, revisionId);
+
+    const relevantMappings = [...allModelMappings.values()].filter((mapping) =>
```

Comment:
Conversion to array doesn't look very efficient here 🙂. I'd suggest testing it on at least 100-200k mappings and see whether it triggers any performance issues.

Reply:
I'm not too worried about this - Fetching the assets to begin with would be much worse performance-wise. It's a linear search through an array. But anyway, I don't know why I'm doing this `[...list.values()]`-thing at all. It's just a list to begin with :thinking: I'll simplify it.